### DISCUSSION TITLE
[Maybe bug]TVM使用变量报错

### DISCUSSION THREAD

**Discussion by hbsun2113** [mlc-ai/mlc-zh#33] [Q&A]
在这个[练习](https://mlc.ai/chapter_tensor_program/tensorir_exercises.html#:~:text=Please%20complete%20the%20following%20Module%20MyConv,%2C%20rtol%3D1e%2D5)中，直接使用整数8,8就没有问题，如果使用变量就会报错
![image](https://user-images.githubusercontent.com/27994442/177543999-3832e7f2-e2d5-4592-b627-7f0fb5ca88f3.png)

---------------------------------------------------------------------------
DiagnosticError                           Traceback (most recent call last)
[<ipython-input-58-8f299127363b>](https://localhost:8080/#) in <module>()
     25         C[vb, vk, vi, vj] += A[vb, vq, vi+vdi, vj+vdj] * B[vk, vq, vdi, vdj]
     26 
---> 27 @tvm.script.ir_module
     28 class MyConv2:
     29   @T.prim_func

20 frames
[<ipython-input-58-8f299127363b>](https://localhost:8080/#) in MyConv2()
     28 class MyConv2:
     29   @T.prim_func
---> 30   def conv(A: T.Buffer[(N, CI, H, W), "int64"], B: T.Buffer[(CO, CI, K, K), "int64"], C: T.Buffer[(N, CO, OUT_H, OUT_W), "int64"]):
     31     T.func_attr({"global_symbol": "conv", "tir.noalias": True})
     32     for b, k, i, j, q, di, dj in T.grid(N, CO, OUT_H, OUT_W, CI, K, K):

[/usr/local/lib/python3.7/dist-packages/tvm/script/tir/prim_func.py](https://localhost:8080/#) in prim_func(input_func)
     38     """
     39     if inspect.isfunction(input_func):
---> 40         result = from_source(input_func)
     41         result.__name__ = input_func.__name__
     42         result.__qualname__ = input_func.__qualname__

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in from_source(input_func, tir_prefix)
   1345         closure_vars = {**_closure_vars.nonlocals, **_closure_vars.globals}
   1346         parser = TVMScriptParser(start_line, namespace, closure_vars)
-> 1347         result = to_ast(input_func, TVMDiagnosticCtx(), parser)
   1348         return result
   1349     else:

[/usr/local/lib/python3.7/dist-packages/synr/compiler.py](https://localhost:8080/#) in to_ast(program, diagnostic_ctx, transformer)
    779         return err
    780     if transformer is not None:
--> 781         transformed = transformer.do_transform(prog, diagnostic_ctx)
    782         err = diagnostic_ctx.render()
    783         if err is not None:

[/usr/local/lib/python3.7/dist-packages/synr/transformer.py](https://localhost:8080/#) in do_transform(self, node, diag)
     28         """
     29         self._diagnostic_context = diag
---> 30         return self.transform(node)
     31 
     32     def error(self, message, span):

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Module(self, node)
    362         """
    363         if len(node.funcs) == 1:
--> 364             return self.transform(next(iter(node.funcs.values())))
    365         elif len(node.func) == 0:
    366             self.report_error(

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Function(self, node)
    458                 self.transform(arg.ty.func_name), ty.GenericBufferType
    459             ):
--> 460                 result = self.handle_match_buffer_type(arg.ty, arg.name)
    461                 if not isinstance(result, buffer.Buffer):
    462                     self.report_error(

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in handle_match_buffer_type(self, node, buffer_name)
   1289         self._inside_buffer_sugar = True
   1290         try:
-> 1291             arg_list = self.parse_arg_list(func, node)
   1292         finally:
   1293             self._inside_buffer_sugar = False

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_arg_list(self, func, node_call)
    270         assert isinstance(node_call, (ast.Call, ast.TypeApply, ast.TypeCall))
    271         # collect arguments
--> 272         args = [self.transform(arg) for arg in node_call.params]
    273         if isinstance(node_call, ast.TypeApply):
    274             kw_args = {}  # TypeApply (e.g. foo[bar]) doesn't have kwargs defined in synr

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in <listcomp>(.0)
    270         assert isinstance(node_call, (ast.Call, ast.TypeApply, ast.TypeCall))
    271         # collect arguments
--> 272         args = [self.transform(arg) for arg in node_call.params]
    273         if isinstance(node_call, ast.TypeApply):
    274             kw_args = {}  # TypeApply (e.g. foo[bar]) doesn't have kwargs defined in synr

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_TypeTuple(self, node)
   1229         Mostly used in `transform_TypeCall` and `transform_TypeApply`.
   1230         """
-> 1231         return [self.transform(value) for value in node.values]
   1232 
   1233     def transform_TypeCall(self, node):

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in <listcomp>(.0)
   1229         Mostly used in `transform_TypeCall` and `transform_TypeApply`.
   1230         """
-> 1231         return [self.transform(value) for value in node.values]
   1232 
   1233     def transform_TypeCall(self, node):

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Var(self, node)
   1193         if symbol is not None:
   1194             return symbol
-> 1195         self.report_error(f"Unknown identifier {name}.", node.span)
   1196 
   1197     def transform_TypeVar(self, node):

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in report_error(self, message, span)
    218         if isinstance(span, tvm.ir.Span):
    219             span = synr_span_from_tvm(span)
--> 220         self.error(message, span)
    221 
    222     def parse_body(self, parent):

[/usr/local/lib/python3.7/dist-packages/synr/transformer.py](https://localhost:8080/#) in error(self, message, span)
     32     def error(self, message, span):
     33         """Report an error on a given span."""
---> 34         self._diagnostic_context.emit("error", message, span)
     35 
     36     def transform(self, node: ast.Node) -> Union[M, F, S, E, B, P, T, None]:

[/usr/local/lib/python3.7/dist-packages/tvm/script/diagnostics.py](https://localhost:8080/#) in emit(self, _level, message, span)
     50         span = self.to_tvm_span(self.source_name, span)
     51         self.diag_ctx.emit(Diagnostic(DiagnosticLevel.ERROR, span, message))
---> 52         self.diag_ctx.render()  # Raise exception on the first error we hit. TODO remove
     53 
     54     def render(self):

[/usr/local/lib/python3.7/dist-packages/tvm/ir/diagnostics/__init__.py](https://localhost:8080/#) in render(self)
    117     def render(self):
    118         """Render the current context using its renderer member."""
--> 119         _ffi_api.DiagnosticContextRender(self)

tvm/_ffi/_cython/./packed_func.pxi in tvm._ffi._cy3.core.PackedFuncBase.__call__()

tvm/_ffi/_cython/./packed_func.pxi in tvm._ffi._cy3.core.FuncCall()

tvm/_ffi/_cython/./packed_func.pxi in tvm._ffi._cy3.core.FuncCall3()

tvm/_ffi/_cython/./base.pxi in tvm._ffi._cy3.core.CHECK_CALL()

DiagnosticError: Traceback (most recent call last):
  3: TVMFuncCall
  2: tvm::runtime::PackedFuncObj::Extractor<tvm::runtime::PackedFuncSubObj<tvm::runtime::TypedPackedFunc<void (tvm::DiagnosticContext)>::AssignTypedLambda<tvm::__mk_TVM11::{lambda(tvm::DiagnosticContext)#1}>(tvm::__mk_TVM11::{lambda(tvm::DiagnosticContext)#1}, std::string)::{lambda(tvm::runtime::TVMArgs const&, tvm::runtime::TVMRetValue*)#1}> >::Call(tvm::runtime::PackedFuncObj const*, tvm::__mk_TVM11::{lambda(tvm::DiagnosticContext)#1}, tvm::runtime::TVMArgs const&)
  1: tvm::DiagnosticContext::Render()
  0: _ZN3tvm7runtime6deta
  File "/workspace/tvm/src/ir/diagnostic.cc", line 105
DiagnosticError: one or more error diagnostics were emitted, please check diagnostic render for output.

**Comment by hbsun2113** [comment]
有人遇到相同问题：https://github.com/mlc-ai/mlc-zh/discussions/28

**Reply by elinx** [reply]
怎么看出来是H和W的问题的？这一大段报错太劝退了

**Reply by Hzfengsy** [reply]
目前来说不建议在TVMScript中使用Python变量，对于固定shape建议使用常数。相关功能社区正在实现中

**Comment by un-knight** [comment]
我遇到了类似的问题，但我的问题更加严重，我将Buffer type hint中维度标注写成**变量或者常数**的形式都会遇到报错。
![image](https://user-images.githubusercontent.com/8266614/178399581-fb5e8a87-5bcf-4501-83f9-113bb052ef85.png)

![image](https://user-images.githubusercontent.com/8266614/178399935-2403eb3e-0151-49ac-8e97-906e58d3673f.png)


错误信息：
```
---------------------------------------------------------------------------
DiagnosticError                           Traceback (most recent call last)
[<ipython-input-29-1e08d9c65db1>](https://localhost:8080/#) in <module>()
      2 OUT_H, OUT_W = H-K+1, W-K+1
      3 
----> 4 @tvm.script.ir_module
      5 class MyConv:
      6   @T.prim_func

42 frames
[<ipython-input-29-1e08d9c65db1>](https://localhost:8080/#) in MyConv()
      7   def conv(inp: T.Buffer[(N, CI, H, W), "int64"], 
      8       weight: T.Buffer[(CO, CI, K, K), "int64"],
----> 9       oup: T.Buffer[(N, CO, OUT_H, OUT_W), "int64"]):
     10     T.func_attr({"global_symbol": "conv", "tir.noalias": True})
     11     for n, co, out_h, out_w, ci, kh, kw in T.grid(N, CO, OUT_H, OUT_W, CI, K, K):

[/usr/local/lib/python3.7/dist-packages/tvm/script/tir/prim_func.py](https://localhost:8080/#) in prim_func(input_func)
     38     """
     39     if inspect.isfunction(input_func):
---> 40         result = from_source(input_func)
     41         result.__name__ = input_func.__name__
     42         result.__qualname__ = input_func.__qualname__

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in from_source(input_func, tir_prefix)
   1345         closure_vars = {**_closure_vars.nonlocals, **_closure_vars.globals}
   1346         parser = TVMScriptParser(start_line, namespace, closure_vars)
-> 1347         result = to_ast(input_func, TVMDiagnosticCtx(), parser)
   1348         return result
   1349     else:

[/usr/local/lib/python3.7/dist-packages/synr/compiler.py](https://localhost:8080/#) in to_ast(program, diagnostic_ctx, transformer)
    779         return err
    780     if transformer is not None:
--> 781         transformed = transformer.do_transform(prog, diagnostic_ctx)
    782         err = diagnostic_ctx.render()
    783         if err is not None:

[/usr/local/lib/python3.7/dist-packages/synr/transformer.py](https://localhost:8080/#) in do_transform(self, node, diag)
     28         """
     29         self._diagnostic_context = diag
---> 30         return self.transform(node)
     31 
     32     def error(self, message, span):

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Module(self, node)
    362         """
    363         if len(node.funcs) == 1:
--> 364             return self.transform(next(iter(node.funcs.values())))
    365         elif len(node.func) == 0:
    366             self.report_error(

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Function(self, node)
    481 
    482         # fetch the body of root block
--> 483         body = self.parse_body(node.body)
    484 
    485         # return a tir.PrimFunc

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_body(self, parent)
    234             stmt = self.context.node_stack[-1].pop()
    235             spans.append(stmt.span)
--> 236             res = self.transform(stmt)
    237             if res is not None:
    238                 body.append(res)

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_For(self, node)
    762         ]
    763         func.enter_scope(node, self.context, arg_list, node.rhs.func_name.span)
--> 764         func.body = self.parse_body(node)
    765         res = func.exit_scope(node, self.context, arg_list, node.rhs.func_name.span)
    766         # exit the scope

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_body(self, parent)
    234             stmt = self.context.node_stack[-1].pop()
    235             spans.append(stmt.span)
--> 236             res = self.transform(stmt)
    237             if res is not None:
    238                 body.append(res)

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_With(self, node)
    814         arg_list = self.parse_arg_list(func, node.rhs)
    815         func.enter_scope(node, self.context, arg_list, node.rhs.func_name.span)
--> 816         func.body = self.parse_body(node)
    817         res = func.exit_scope(node, self.context, arg_list, node.rhs.func_name.span)
    818         # exit the scope

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_body(self, parent)
    234             stmt = self.context.node_stack[-1].pop()
    235             spans.append(stmt.span)
--> 236             res = self.transform(stmt)
    237             if res is not None:
    238                 body.append(res)

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Assign(self, node)
    583                 arg_list = self.parse_arg_list(func, node.rhs)
    584                 func.handle(node, self.context, arg_list, node.rhs.func_name.span)
--> 585                 return self.parse_body(node)
    586             elif isinstance(func, types.FunctionType):
    587                 # Pattern 5

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_body(self, parent)
    234             stmt = self.context.node_stack[-1].pop()
    235             spans.append(stmt.span)
--> 236             res = self.transform(stmt)
    237             if res is not None:
    238                 body.append(res)

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Assign(self, node)
    583                 arg_list = self.parse_arg_list(func, node.rhs)
    584                 func.handle(node, self.context, arg_list, node.rhs.func_name.span)
--> 585                 return self.parse_body(node)
    586             elif isinstance(func, types.FunctionType):
    587                 # Pattern 5

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_body(self, parent)
    234             stmt = self.context.node_stack[-1].pop()
    235             spans.append(stmt.span)
--> 236             res = self.transform(stmt)
    237             if res is not None:
    238                 body.append(res)

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Assign(self, node)
    583                 arg_list = self.parse_arg_list(func, node.rhs)
    584                 func.handle(node, self.context, arg_list, node.rhs.func_name.span)
--> 585                 return self.parse_body(node)
    586             elif isinstance(func, types.FunctionType):
    587                 # Pattern 5

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_body(self, parent)
    234             stmt = self.context.node_stack[-1].pop()
    235             spans.append(stmt.span)
--> 236             res = self.transform(stmt)
    237             if res is not None:
    238                 body.append(res)

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Assign(self, node)
    583                 arg_list = self.parse_arg_list(func, node.rhs)
    584                 func.handle(node, self.context, arg_list, node.rhs.func_name.span)
--> 585                 return self.parse_body(node)
    586             elif isinstance(func, types.FunctionType):
    587                 # Pattern 5

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_body(self, parent)
    234             stmt = self.context.node_stack[-1].pop()
    235             spans.append(stmt.span)
--> 236             res = self.transform(stmt)
    237             if res is not None:
    238                 body.append(res)

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Assign(self, node)
    583                 arg_list = self.parse_arg_list(func, node.rhs)
    584                 func.handle(node, self.context, arg_list, node.rhs.func_name.span)
--> 585                 return self.parse_body(node)
    586             elif isinstance(func, types.FunctionType):
    587                 # Pattern 5

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_body(self, parent)
    234             stmt = self.context.node_stack[-1].pop()
    235             spans.append(stmt.span)
--> 236             res = self.transform(stmt)
    237             if res is not None:
    238                 body.append(res)

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Assign(self, node)
    581             elif isinstance(func, SpecialStmt):
    582                 # Pattern 1
--> 583                 arg_list = self.parse_arg_list(func, node.rhs)
    584                 func.handle(node, self.context, arg_list, node.rhs.func_name.span)
    585                 return self.parse_body(node)

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in parse_arg_list(self, func, node_call)
    270         assert isinstance(node_call, (ast.Call, ast.TypeApply, ast.TypeCall))
    271         # collect arguments
--> 272         args = [self.transform(arg) for arg in node_call.params]
    273         if isinstance(node_call, ast.TypeApply):
    274             kw_args = {}  # TypeApply (e.g. foo[bar]) doesn't have kwargs defined in synr

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in <listcomp>(.0)
    270         assert isinstance(node_call, (ast.Call, ast.TypeApply, ast.TypeCall))
    271         # collect arguments
--> 272         args = [self.transform(arg) for arg in node_call.params]
    273         if isinstance(node_call, ast.TypeApply):
    274             kw_args = {}  # TypeApply (e.g. foo[bar]) doesn't have kwargs defined in synr

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Call(self, node)
    860                 return self.transform_Subscript(node)
    861             if node.func_name.name in self._binop_maker:
--> 862                 lhs = self.transform(node.params[0])
    863                 # There is no supertype for everything that can appear in
    864                 # an expression, so we manually add what we might get here.

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform(self, node)
    194         method = "transform_" + node.__class__.__name__
    195         visitor = getattr(self, method, self.generic_visit)
--> 196         transform_res = visitor(node)
    197 
    198         self.current_lineno, self.current_col_offset = old_lineno, old_col_offset

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in transform_Call(self, node)
    871                         f"Left hand side of binary op must be a PrimExpr, "
    872                         "but it is a {type(lhs).__name__}",
--> 873                         node.params[0].span,
    874                     )
    875                 rhs = self.transform(node.params[1])

[/usr/local/lib/python3.7/dist-packages/tvm/script/parser.py](https://localhost:8080/#) in report_error(self, message, span)
    218         if isinstance(span, tvm.ir.Span):
    219             span = synr_span_from_tvm(span)
--> 220         self.error(message, span)
    221 
    222     def parse_body(self, parent):

[/usr/local/lib/python3.7/dist-packages/synr/transformer.py](https://localhost:8080/#) in error(self, message, span)
     32     def error(self, message, span):
     33         """Report an error on a given span."""
---> 34         self._diagnostic_context.emit("error", message, span)
     35 
     36     def transform(self, node: ast.Node) -> Union[M, F, S, E, B, P, T, None]:

[/usr/local/lib/python3.7/dist-packages/tvm/script/diagnostics.py](https://localhost:8080/#) in emit(self, _level, message, span)
     50         span = self.to_tvm_span(self.source_name, span)
     51         self.diag_ctx.emit(Diagnostic(DiagnosticLevel.ERROR, span, message))
---> 52         self.diag_ctx.render()  # Raise exception on the first error we hit. TODO remove
     53 
     54     def render(self):

[/usr/local/lib/python3.7/dist-packages/tvm/ir/diagnostics/__init__.py](https://localhost:8080/#) in render(self)
    117     def render(self):
    118         """Render the current context using its renderer member."""
--> 119         _ffi_api.DiagnosticContextRender(self)

tvm/_ffi/_cython/./packed_func.pxi in tvm._ffi._cy3.core.PackedFuncBase.__call__()

tvm/_ffi/_cython/./packed_func.pxi in tvm._ffi._cy3.core.FuncCall()

tvm/_ffi/_cython/./packed_func.pxi in tvm._ffi._cy3.core.FuncCall3()

tvm/_ffi/_cython/./base.pxi in tvm._ffi._cy3.core.CHECK_CALL()

DiagnosticError: Traceback (most recent call last):
  3: TVMFuncCall
  2: tvm::runtime::PackedFuncObj::Extractor<tvm::runtime::PackedFuncSubObj<tvm::runtime::TypedPackedFunc<void (tvm::DiagnosticContext)>::AssignTypedLambda<tvm::__mk_TVM11::{lambda(tvm::DiagnosticContext)#1}>(tvm::__mk_TVM11::{lambda(tvm::DiagnosticContext)#1}, std::string)::{lambda(tvm::runtime::TVMArgs const&, tvm::runtime::TVMRetValue*)#1}> >::Call(tvm::runtime::PackedFuncObj const*, tvm::__mk_TVM11::{lambda(tvm::DiagnosticContext)#1}, tvm::runtime::TVMArgs const&)
  1: tvm::DiagnosticContext::Render()
  0: _ZN3tvm7runtime6deta
  File "/workspace/tvm/src/ir/diagnostic.cc", line 105
DiagnosticError: one or more error diagnostics were emitted, please check diagnostic render for output.
```

不知道有什么办法可以解决吗？所用IDE是colab。 @Hzfengsy

**Reply by un-knight** [reply]
很奇怪的一个事情时，当我在vscode IDE上用相同代码执行的时候，却没有以上错误。所以我猜这可能是TVM的某种bug。 @Hzfengsy

**Reply by Hzfengsy** [reply]
感谢提供信息，应该是TVMScript针对IPython的支持问题，我们会进行检查和修复。
